src/lightning/fabric/plugins/precision/bitsandbytes.py

```diff
@@ -192,51 +221,160 @@ class _Linear8bitLt(bnb.nn.Linear8bitLt):
 
         def __init__(self, *args: Any, device: Optional[_DEVICE] = None, threshold: float = 6.0, **kwargs: Any) -> None:
             super().__init__(*args, device=device, threshold=threshold, **kwargs)
+            self.weight = cast(bnb.nn.Int8Params, self.weight)  # type: ignore[has-type]
+            self.bias = cast(Optional[torch.nn.Parameter], self.bias)  # type: ignore[has-type]
             # if the device is CUDA or we are under a CUDA context manager, quantize the weight here, so we don't end up
             # filling the device memory with float32 weights which could lead to OOM
             if torch.tensor(0, device=device).device.type == "cuda":
-                self._quantize_weight(self.weight.data)
-            self._register_load_state_dict_pre_hook(partial(_quantize_on_load_hook, self._quantize_weight))
+                self.quantize_()
+            self._register_load_state_dict_pre_hook(partial(_quantize_on_load_hook, self.quantize_))
             self.register_load_state_dict_post_hook(_ignore_missing_weights_hook)
 
-        def _quantize_weight(self, weight: torch.Tensor) -> None:
+        def quantize_(self, weight: Optional[torch.Tensor] = None, device: Optional[torch.device] = None) -> None:
+            """Inplace quantize."""
+            if weight is None:
+                weight = self.weight.data
+                if weight.data.type == torch.uint8:
+                    # already quantized
+                    return
+            assert isinstance(self.weight, bnb.nn.Int8Params)
+            self.weight = self.quantize(self.weight, weight, device)
+
+        @staticmethod
+        def quantize(
+            int8params: bnb.nn.Int8Params, weight: torch.Tensor, device: Optional[torch.device]
+        ) -> bnb.nn.Int8Params:
+            device = device or torch.device("cuda")
+            if device.type != "cuda":
+                raise RuntimeError(f"Unexpected device type: {device.type}")
             # https://github.com/TimDettmers/bitsandbytes/blob/0.41.0/bitsandbytes/nn/modules.py#L291-L302
-            B = weight.contiguous().to(device="cuda", dtype=torch.float16)
-            if self.state.has_fp16_weights:
-                self.weight.data = B
+            B = weight.contiguous().to(device=device, dtype=torch.float16)
+            if int8params.has_fp16_weights:
+                int8params.data = B
             else:
                 CB, CBt, SCB, SCBt, coo_tensorB = bnb.functional.double_quant(B)
                 del CBt
                 del SCBt
```

Comment:
```suggestion
                CB, _, SCB, _, coo_tensorB = bnb.functional.double_quant(B)
```

Reply:
This piece is copied verbatim from bitsandbytes. I don't want to risk a memory issue from changing it.